### PR TITLE
Add Rotate() function into Cuda namespace

### PR DESCRIPTION
**Pull request by MatthewMcGonagle**
Added Image_Function_Cuda::Rotate() and __global__ rotateCuda() Cuda Kernel.

Related to issue #267 Add CUDA implementation of Rotate function.

### CODE REVIEW

File: src/cuda/image_function_cuda.cu [hunk 1]
```diff
@@ -239,6 +239,59 @@ namespace
         }
     }
 
+    __global__ void rotateCuda( const uint8_t * in, uint32_t rowSizeIn, uint32_t inHeight, float inXStart, float inYStart, 
+                                uint8_t * out, uint32_t rowSizeOut, uint32_t outHeight, 
+                                float cosAngle, float sinAngle)
+    {
+        uint32_t outX = blockDim.x * blockIdx.x + threadIdx.x,
+                 outY = blockDim.y * blockIdx.y + threadIdx.y;
```
Comment (ihhub) [inline src/cuda/image_function_cuda.cu]: Please use an explicit declaration of variable type in all cases as people might read that this is a single variable.

```cpp
const uint32_t outX = blockDim.x * blockIdx.x + threadIdx.x,
const uint32_t outY = blockDim.y * blockIdx.y + threadIdx.y;
```

Please change the same for below cases as well.

File: src/cuda/image_function_cuda.cu [hunk 2]
```diff
@@ -239,6 +239,59 @@ namespace
         }
     }
 
+    __global__ void rotateCuda( const uint8_t * in, uint32_t rowSizeIn, uint32_t inHeight, float inXStart, float inYStart, 
+                                uint8_t * out, uint32_t rowSizeOut, uint32_t outHeight, 
+                                float cosAngle, float sinAngle)
```
Comment (ihhub) [inline src/cuda/image_function_cuda.cu]: Move this code to previous line. We have a recommended line size as 150 characters so it should be fine.

File: src/cuda/image_function_cuda.cu [hunk 3]
```diff
@@ -239,6 +239,59 @@ namespace
...
+    {
+        uint32_t outX = blockDim.x * blockIdx.x + threadIdx.x,
+                 outY = blockDim.y * blockIdx.y + threadIdx.y;
+       
+        // If this thread isn't for a valid pixel in the output image, then do nothing. 
+
+        if ( outX >= rowSizeOut || outY >= outHeight )
```
Comment (ihhub) [inline src/cuda/image_function_cuda.cu]: Please change this condition to opposite and put the code inside the condition. `return` is not recommended to use in GPU kernels as all threads expect to finish their work at the end of kernel's body. So the code would be like this:
```cpp
if ( outX < rowSizeOut && outY < outHeight ) {
    // your code below
}
```

File: src/cuda/image_function_cuda.cu [hunk 4]
```diff
@@ -239,6 +239,59 @@ namespace
...
+        if ( outX >= rowSizeOut || outY >= outHeight ) 
+            return;
+
+        // Both input coordinates are shifted using the cosAngle, sinAngle, outX, and outY. The shift
+        // comes from inverse rotating the horizontal and vertical iterations over the output.
+
+        // Note that inverse rotation of horizontal motion is ( cos(angle), -sin(angle)),
+        // and the inverse rotation of vertical motion is ( sin(angle), cos(angle) ).
+
+        float exactInX = inXStart + cosAngle * outX + sinAngle * outY,
```
Comment (ihhub) [inline src/cuda/image_function_cuda.cu]: Please use `const` word wherever it is possible in variable initialisation :)

File: src/cuda/image_function_cuda.cu (hunk 1 above)
Comment (MatthewMcGonagle) [reply src/cuda/image_function_cuda.cu]: Okay, done.

File: src/cuda/image_function_cuda.cu (hunk 2 above)
Comment (MatthewMcGonagle) [reply src/cuda/image_function_cuda.cu]: I moved `float cosAngle, float sinAngle)` to line above.

File: src/cuda/image_function_cuda.cu (hunk 4 above)
Comment (MatthewMcGonagle) [reply src/cuda/image_function_cuda.cu]: I used `const` where appropriate. I also removed the pointer variables `inMem` and `outMem` from `rotateCuda()` as they are unnecessary. We can just use `in` and `out` directly.

File: src/cuda/image_function_cuda.cu (hunk 3 above)
Comment (MatthewMcGonagle) [reply src/cuda/image_function_cuda.cu]: Okay, done.

**Pull request merged**